request.go

```diff
@@ -142,6 +141,135 @@ func UnmarshalManyPayload(in io.Reader, t reflect.Type) ([]interface{}, error) {
 	return models, nil
 }
 
+// jsonapiTypeOfModel returns a jsonapi primary type string
+// given a struct type that has typical jsonapi struct tags
+//
+// Example:
+// For this type, "posts" is returned. An error is returned if
+// no properly-formatted "primary" tag is found for jsonapi
+// annotations
+//
+//	type Post struct {
+//	    ID string `jsonapi:"primary,posts"`
+//	}
+func jsonapiTypeOfModel(structModel reflect.Type) (string, error) {
+	for i := 0; i < structModel.NumField(); i++ {
+		fieldType := structModel.Field(i)
+		args, err := getStructTags(fieldType)
+		if err != nil || len(args) < 2 {
+			continue
+		}
+
+		if args[0] == annotationPrimary {
+			return args[1], nil
+		}
+	}
+
+	return "", errors.New("no primary annotation found on model")
+}
+
+// structFieldIndex holds a bit of information about a type found at a struct field index
+type structFieldIndex struct {
+	Type     reflect.Type
+	FieldNum int
+}
+
+// choiceStructMapping reflects on a value that may be a slice
+// of choice type structs or a choice type struct. A choice type
+// struct is a struct comprising of pointers to other jsonapi models,
+// only one of which is populated with a value by the decoder.
+//
+// The specified type is probed and a map is generated that maps the
+// underlying model type (its 'primary' type) to the field number
+// within the choice type struct. This data can then be used to correctly
+// assign each data relationship node to the correct choice type
+// struct field.
+func choiceStructMapping(choice reflect.Type) (result map[string]structFieldIndex, err error) {
```

Comment:
No, this needs to ignore those probing errors because we want to allow different types of structs with different types of fields to be a choice type. I removed the error return value because it's not used